news/2017-05-19-tour-around-weld-3.asciidoc

```diff
@@ -0,0 +1,194 @@
+desc: A summary of CDI 2.0 features coated with Weld and sprinkled with code samples
+tags: [release]
+---
+:linkattrs:
+
+This post briefly describes all the main CDI 2.0 features and elaborates on Weld-specific features we added on top of that.
+It is not intended as a deep-dive but rather to give you the overall idea of what is going on and what can the new release offer.
+So, enough talk, let's get the show on the road!
+
+=== Async Events and Notification Options
+
+Up until now, the only way to send events was to do so synchronously.
+That meant stopping the work of your current thread to instead occupy it with observer resolution and subsequent notification of observer methods.
+Once all that was done, the thread resumed it's work.
+
+With CDI 2.0 there are asynchronous events - a 'fire & forget' way of handling things.
+Here is how to fire such and event and how to observe it:
+
+[ source, java ]
+----
+// nothing new here, plain old Event is used
+@Inject Event<Payload> event;
+
+public void sendAsyncEvent() {
+    // we use a new fireAsync method for asynchronous events
+    CompletionStage<Void> eventResult = event.fireAsync(new Payload()).thenAccept(...); 
+}
+
+public void asyncObserver (@ObservesAsync Payload payload){ … }
+----
+
+Few things to note here - first of all, the return value of `fireAsync` is `CompletionStage` allowing you to chain the work for when the event is done.
+You can follow-up with more tasks, react on exceptional return value and so on.
+Another noticeable detail is that observer which are to be notified have to have `@ObservesAsync` annotation.
+As you might have guessed, an observer can be notified of either synchronous (`@Observes`) or asynchronous (`@ObservesAsync`) events but never both!
+
+Apart from this very basic usage, async events allow you to specify `NotificationOptions` - an option allowing you to tweak the way notification works.
+From CDI perspective, there is currently only one standardized option and that enables usage of custom executors which will be used for notifications.
+But the API was designed in a generic way so that implementation can enrich it with its own options.
+Weld currently offers two more options - timeout and parallel execution - both of which are well described in link:http://docs.jboss.org/weld/reference/3.0.0.Final/en-US/html/events.html#_notification_options[Weld docs, window="_blank"].
+And here is a snippet showing it in action:
+
+[ source, java ]
+----
+public void sendAsyncEventWithOptions() {
+    // we use a secondary version of fireAsync method with NotificationOptions parameter
+    event.fireAsync(new Payload(), NotificationOptions.of("weld.async.notification.timeout", "2000")); 
+}
+----
+
+
+=== Configurators API
+
+Another of the key features are link:http://docs.jboss.org/cdi/spec/2.0/cdi-spec.html#configurators[`Configurators`, window="blank"].
+They are there to ease our work when doing small tweaks in extensions.
```

Comment:
`They are here` sounds better to me... but it's probably just a matter of taste ;-)